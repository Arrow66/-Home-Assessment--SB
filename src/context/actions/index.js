export const FETCH_DATA_SUCCESS = "FETCH_DATA_SUCCESS";
export const SET_SELECTED_ACCOUNT = "SET_SELECTED_ACCOUNT";
export const SET_DEFAULT_ACCOUNT = "SET_DEFAULT_ACCOUNT";


