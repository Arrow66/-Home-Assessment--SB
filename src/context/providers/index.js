import AccountsProvider from "./AccountsProvider";

export default {AccountsProvider};