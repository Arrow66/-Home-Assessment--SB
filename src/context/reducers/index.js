import reducer from "./accountsReducer"

export default reducer