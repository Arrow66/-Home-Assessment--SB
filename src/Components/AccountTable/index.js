import React, { useContext } from "react"
import Table from "../Table"
import { accountsContext } from "../../context";


const { ValueContext } = accountsContext


const AccountTable = () => {
    const { selectedAccount, accounts } = useContext(ValueContext);
    const tableData = accounts[selectedAccount];
    const { docs } = tableData || {}

    const TableRender = docs&&docs.map(({ document_type, is_paper_flag }) => (
        <Table.Row>
            <Table.Cell>{document_type}</Table.Cell>
            <Table.Cell>{is_paper_flag ? "Paper" : "Paperless"}</Table.Cell>
        </Table.Row>
    ))||[]

    return (
        <Table>
            <Table.Header>
                <Table.Column> Document Type</Table.Column>
                <Table.Column>Delivery Method</Table.Column>
            </Table.Header>
            <Table.Body>
                {
                    TableRender
                }
            </Table.Body>
        </Table>
    )

}

export default AccountTable;