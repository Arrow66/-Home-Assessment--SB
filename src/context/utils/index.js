import { FETCH_DATA_SUCCESS,SET_DEFAULT_ACCOUNT,SET_SELECTED_ACCOUNT } from "../actions";


export const fetchDataSuccess = async (dispatch, data) => {

    const { accounts, prefDocuments } = data;
    const finalData = accounts.reduce((acc, account) => {
        const documentbyAccountKey = prefDocuments.find((document) => {
            return document.account_key === account.key;
        })
        acc[account.key] = {
            accountName: account.description + " " + (account.typeDesc !== "IMA" ? account.id.slice(0, 3) + "-" + account.id.slice(3) : account.id) + " " + account.typeDesc,
            docs: documentbyAccountKey.docs
        };

        return acc
    }, {})

    dispatch({ type: FETCH_DATA_SUCCESS, payload: { finalData } })

}



export const setSelectedAccountAction = (dispatch,key)=>{
    dispatch({
       type: SET_SELECTED_ACCOUNT,
       payload:key
    })
}

export const setDefaultAccount  = (dispatch,defaultAccount)=>{
    const {data} = defaultAccount;
    dispatch({
       type: SET_SELECTED_ACCOUNT,
       payload:data.default_accounts[0].account_key
    })
}