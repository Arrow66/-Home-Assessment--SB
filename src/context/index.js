//contexts
import { ValueContext, ValueDispatchContext } from "./providers/AccountsProvider";



export const accountsContext = {
    ValueContext,
    ValueDispatchContext
}

