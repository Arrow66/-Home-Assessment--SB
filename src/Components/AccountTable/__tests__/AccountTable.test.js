import React from 'react';
import { shallow, mount } from 'enzyme';
import { accountsContext } from '../../../context';
import AccountTable from '..';
import renderer from "react-test-renderer"


describe('AccountTable', () => {
  let wrapper;
  const mockAccounts = {
    'account1': {
      docs: [
        { document_type: 'doc1', is_paper_flag: true },
        { document_type: 'doc2', is_paper_flag: false },
      ],
    },
    'account2': {
      docs: [
        { document_type: 'doc3', is_paper_flag: true },
        { document_type: 'doc4', is_paper_flag: false },
      ],
    },
  };
const {ValueContext} = accountsContext
  const mockContext = {
    selectedAccount: 'account1',
    accounts: mockAccounts,
  };

  beforeEach(() => {
    wrapper = mount(
      <ValueContext.Provider value={mockContext}>
        <AccountTable />
      </ValueContext.Provider>
    );
  });

  it('should render Table component with data-test attribute', () => {
    expect(wrapper.find({"data-testid": "account-table"}).exists()).toBeTruthy();
  });

  it('should render Table.Header component with data-test attribute', () => {
    expect(wrapper.find({"data-testid":"account-table-header"}).exists()).toBeTruthy();
  });

  it('should render Table.Body component with data-test attribute', () => {
    expect(wrapper.find({"data-testid":"account-table-body"}).exists()).toBeTruthy();
  });

  it('should render correct number of Table.Row components with data-test attribute', () => {
    expect(wrapper.find('Table.Row[data-testid="account-table-row"]')).toBeTruthy();
  });

  it('should render correct document types and delivery methods with data-test attribute', () => {
    const rows = wrapper.find({'data-testid':"account-table-row"}).at(0);
    expect(rows.at(0).find({"data-testid":"document_type"}).at(0).text()).toEqual('doc1');
    expect(rows.at(0).find({"data-testid":"delivery-method"}).at(0).text()).toEqual('Paper');
  });

  it('should render empty Table.Body when no data is available', () => {
    const emptyContext = {
      selectedAccount: 'account3',
      accounts: mockAccounts,
    };
    wrapper = mount(
      <ValueContext.Provider value={emptyContext}>
        <AccountTable />
      </ValueContext.Provider>
    );
    expect(wrapper.find('Table.Row[data-testid="account-table-row"]')).toHaveLength(0);
  });

it("should match snap",()=>{
  const tree =   renderer.create(
        <ValueContext.Provider value={mockContext}>
          <AccountTable />
        </ValueContext.Provider>
      ).toJSON();
      expect(tree).toMatchSnapshot();
})



});
