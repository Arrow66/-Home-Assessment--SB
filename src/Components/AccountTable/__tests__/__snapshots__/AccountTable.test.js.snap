// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountTable should match snap 1`] = `
<table
  aria-label="Account Table"
  className="sc-beySPh dwkWpT"
  data-testid="account-table"
  role="table"
>
  <thead
    className="sc-eDLKkx jQvHwW"
    data-testid="account-table-header"
    role="rowheader"
  >
    <tr
      className="sc-hLQSwg fYZOnz"
      tabIndex={-1}
    >
      <th
        className="sc-dmyCSP cjlLNF"
        role="columnheader"
      >
         Document Type
      </th>
      <th
        className="sc-dmyCSP cjlLNF"
        role="columnheader"
      >
        Delivery Method
      </th>
    </tr>
  </thead>
  <tbody
    aria-live="polite"
    className="sc-guDLey cVtAAz"
    data-testid="account-table-body"
    role="rowgroup"
  >
    <tr
      className="sc-hLQSwg fYZOnz"
      data-testid="account-table-row"
      role="row"
      tabIndex={-1}
    >
      <td
        align="left"
        className="sc-jTQCzO iZOOrt"
        data-testid="document_type"
        role="cell"
      >
        doc1
      </td>
      <td
        align="left"
        className="sc-jTQCzO iZOOrt"
        data-testid="delivery-method"
        role="cell"
      >
        Paper
      </td>
    </tr>
    <tr
      className="sc-hLQSwg fYZOnz"
      data-testid="account-table-row"
      role="row"
      tabIndex={-1}
    >
      <td
        align="left"
        className="sc-jTQCzO iZOOrt"
        data-testid="document_type"
        role="cell"
      >
        doc2
      </td>
      <td
        align="left"
        className="sc-jTQCzO iZOOrt"
        data-testid="delivery-method"
        role="cell"
      >
        Paperless
      </td>
    </tr>
  </tbody>
</table>
`;
